cli/resendfailed.php

```diff
@@ -0,0 +1,173 @@
+
+
+// Start of sanitization, and applying of scope.
+$options['dryrun'] = boolval($options['dryrun']);
+if ($options['dryrun']) {
+    cli_write('NOTICE: The program is running in dryrun mode, no write queries will be executed. ');
+    cli_write('To disable dryrun mode, add option --dryrun=0' . PHP_EOL);
+}
+
+$options['batch'] = intval($options['batch']);
+cli_writeln("Program will run in batches of {$options['batch']} records ...");
+
+if (!empty($options['errortype'])) {
+    $options['errortype'] = explode(',', $options['errortype']);
+    $options['errortype'] = array_map('intval', $options['errortype']);
+    list($insql, $inparams) = $DB->get_in_or_equal($options['errortype'], SQL_PARAMS_NAMED, 'errt');
+    $where[] = "x.errortype $insql";
+    $params = array_merge($params, $inparams);
+    cli_writeln('Applied scope for errortype ...');
+}
+
+if (!empty($options['eventname'])) {
+    $options['eventname'] = explode(',', $options['eventname']);
+    list($insql, $inparams) = $DB->get_in_or_equal($options['eventname'], SQL_PARAMS_NAMED, 'evt');
+    $where[] = "x.eventname $insql";
+    $params = array_merge($params, $inparams);
+    cli_writeln('Applied scope for eventname ...');
+}
+
+if (!is_null($options['datefrom'])) {
+    $options['datefrom'] = intval($options['datefrom']);
+    $where[] = 'x.timecreated >= :datefrom';
+    $params['datefrom'] = $options['datefrom'];
+    cli_writeln('Applied scope for datefrom ...');
+}
+
+if (!is_null($options['dateto'])) {
+    $options['dateto'] = intval($options['dateto']);
+    $where[] = 'x.timecreated <= :dateto';
+    $params['dateto'] = $options['dateto'];
+    cli_writeln('Applied scope for dateto ...');
+}
+
+if (!is_null($options['datefrom']) && !is_null($options['dateto'])) {
+    if ($options['datefrom'] > $options['dateto']) {
+        cli_writeln(get_string('datetovalidation', 'logstore_xapi'));
+        exit(2);
+    }
+}
+
+if (empty($where)) {
+    $where[] = '1 = 1';
+    cli_writeln('No scope applied, moving all records ...');
+}
+
+$where = implode(' AND ', $where);
+
+$sql = "SELECT x.id
+          FROM {{$basetable}} x
+         WHERE $where";
+
+$limitfrom = 0;
+$limitnum = $options['batch'];
+$counttotal = 0;
+$countsucc = 0;
+$countfail = 0;
+
+do {
+    $records = $DB->get_records_sql($sql, $params, $limitfrom, $limitnum);
+    $count = count($records);
+    $limitfrom += $count;
+    $counttotal += $count;
+    cli_writeln("Read {$count} records, next offset will be {$limitfrom}");
+
+    $eventids = array_keys($records);
+
+    if (empty($eventids)) {
+        continue;
+    }
+
+    $mover = new \logstore_xapi\log\moveback($eventids, XAPI_REPORT_ID_ERROR);
+
+    if ($options['dryrun']) {
+        continue;
+    }
+
+    if ($mover->execute()) {
+        $countsucc += $count;
+        cli_writeln("$count events successfully sent for reprocessing.");
+    } else {
+        $countfail += $count;
+        cli_writeln("$count events failed to sent for reprocessing.");
```

Comment:
Small grammatical error, "send" instead of "sent" would make more sense here, or "... failed to be sent ...".

Reply:
Thanks, will amend.